speaker/bgp_controller.go

```diff
@@ -272,20 +447,301 @@ type session interface {
+		peers := []peer{}
+		for _, p := range c.peers {
+			cp := peer{}
+
+			cp.BGP = p.BGP
+			cfgCopy := *p.Cfg
+			cp.Cfg = &cfgCopy
+
+			cp.Cfg.Password = "REDACTED"
+
+			peers = append(peers, cp)
+		}
+
+		var np peer
+		if c.nodePeer != nil {
+			np = peer{}
+			np.BGP = c.nodePeer.BGP
+			cfgCopy := *c.nodePeer.Cfg
+			np.Cfg = &cfgCopy
+
+			np.Cfg.Password = "REDACTED"
+		}
+
+		res := struct {
+			Peers    []peer
+			NodePeer peer
+		}{
+			Peers:    peers,
+			NodePeer: np,
+		}
+
+		j, err := json.MarshalIndent(res, "", "  ")
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to get status: %s", err), 500)
+			return
+		}
+		fmt.Fprint(w, string(j))
+	}
+}
+
+// parseNodePeer attempts to construct a BGP peer from information conveyed
+// in node annotations and labels using the specified autodiscovery
+// configuration.
+func parseNodePeer(l log.Logger, pad *config.PeerAutodiscovery, anns labels.Set, ls labels.Set) (*peer, error) {
+	var (
+		myASN       uint32
+		peerASN     uint32
+		peerAddr    net.IP
+		peerPort    uint16
+		srcAddr     net.IP
+		holdTime    time.Duration
+		holdTimeRaw string
+		routerID    net.IP
+	)
+
+	// Hardcoded defaults. Used only if a parameter isn't specified in the peer
+	// autodiscovery defaults and also not in annotations/labels.
+	peerPort = 179
+	holdTime = 90 * time.Second
+
+	// Method called with a nil or empty peer autodiscovery.
+	if pad == nil {
+		return nil, errors.New("nil peer autodiscovery")
+	}
+
+	// Set defaults. Parameter values read from labels/annotations override the
+	// values set here.
+	if pad.Defaults != nil {
+		if pad.Defaults.ASN != 0 {
+			peerASN = pad.Defaults.ASN
+		}
+		if pad.Defaults.MyASN != 0 {
+			myASN = pad.Defaults.MyASN
+		}
+		if pad.Defaults.Address != nil {
+			peerAddr = pad.Defaults.Address
+		}
+		if pad.Defaults.Port != 0 {
+			peerPort = pad.Defaults.Port
+		}
+		if pad.Defaults.HoldTime != 0 {
+			holdTime = pad.Defaults.HoldTime
+		}
+	}
+
+	if pad.FromLabels != nil {
+		for k, v := range ls {
+			switch k {
+			case pad.FromLabels.MyASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing local ASN: %v", err)
+				}
+				myASN = uint32(asn)
+			case pad.FromLabels.ASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer ASN: %v", err)
+				}
+				peerASN = uint32(asn)
+			case pad.FromLabels.Addr:
+				peerAddr = net.ParseIP(v)
+				if peerAddr == nil {
+					return nil, fmt.Errorf("invalid peer IP %q", v)
+				}
+			case pad.FromLabels.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromLabels.SrcAddr:
+				srcAddr = net.ParseIP(v)
+				if v != "" && srcAddr == nil {
+					return nil, fmt.Errorf("invalid source IP %q", v)
+				}
+			case pad.FromLabels.HoldTime:
+				holdTimeRaw = v
```

Comment:
All other values were validated in this switch, but the validation for just this one field happens afterwards.  Maybe move the validation up into the switch statements?  It duplicates the validations between the label and annotation switch statements, but that's done for every other field already ...